File: leetcode/addTwoNumbers.js
// https://leetcode.com/problems/add-two-numbers/
//
// You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order, and each of their nodes contains a single digit. Add the two numbers and return the sum as a linked list.
//
// You may assume the two numbers do not contain any leading zero, except the number 0 itself.
//
// Example 1:
// Input: l1 = [2,4,3], l2 = [5,6,4]
// Output: [7,0,8]
// Explanation: 342 + 465 = 807.


/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
let addTwoNumbers = function (l1, l2) {

    let s1 = [];
    let newl1 = l1;

    while (newl1 !== null) {
        s1.unshift(newl1.val.toString());
        newl1 = newl1.next;
    }

    var n1 = "";
    for (let i = 0; i < s1.length; i++) {
        const element = s1[i];
        n1 += element;
    }


    let s2 = [];
    let newl2 = l2;

    while (newl2 !== null) {
        s2.unshift(newl2.val.toString());
        newl2 = newl2.next;
    }

    var n2 = "";
    for (let i = 0; i < s2.length; i++) {
        const element = s2[i];
        n2 += element;
    }
    

    let numberAddition = BigInt(n1) + BigInt(n2);
    let addition = Array.from((numberAddition).toLocaleString('fullwide', {useGrouping:false}), Number);
    
    let result;

    for (let i = 0; i < addition.length; i++) {
        const element = addition[i];
        result = new ListNode(element, result);
    }

    return result
};

// Runtime: 3197 ms, faster than 5.01% of JavaScript online submissions for Add Two Numbers.
// Memory Usage: 56.2 MB, less than 5.06% of JavaScript online submissions for Add Two Numbers.

module.exports = {
    addTwoNumbers
}

File: leetcode/algorithms/Algorithms/2-Add-Two-Numbers/AddTwoNumbers.cs
/*
    Source: https://leetcode.com/problems/add-two-numbers
    Runtime: 116 ms, faster than 62.82% of C# online submissions for Add Two Numbers.
    Memory Usage: 49.4 MB, less than 9.75% of C# online submissions for Add Two Numbers.
    Author: berkslv
    Date: 14/06/2022
*/
using Algorithms.Utils;

namespace Algorithms
{
    public class Solution002
    {
        public ListNode AddTwoNumbers(ListNode l1, ListNode l2)
        {
            var head = new ListNode(0);
            int n1 = 0;
            int n2 = 0;
            int carry = 0;
            var current = head;

            while (l1 != null || l2 != null || carry != 0)
            {

                if (l1 == null)
                {
                    n1 = 0;
                } else
                {
                    n1 = l1.val;
                    l1 = l1.next;
                }

                if (l2 == null)
                {
                    n2 = 0;
                } else
                {
                    n2 = l2.val;
                    l2 = l2.next;
                }

                current.next = new ListNode((n1 + n2 + carry) % 10);
                current = current.next;
                carry = (n1 + n2 + carry) / 10;
            }

            return head.next;
        }
    }
}
